lilya/_internal/_encoders2.py

```diff
@@ -0,0 +1,272 @@
+    def serialize(self, obj: PurePath) -> str:
+        return str(obj)
+
+    def encode(self, structure: type[PurePath], value: Any) -> Enum:
+        return structure(value)
+
+
+class DateEncoder(Encoder[Any], WithEncodeProtocol):
+    name: str = "DateEncoder"
+    __type__ = date
+
+    def serialize(self, obj: date) -> str:
+        return obj.isoformat()
+
+    def encode(self, structure: type[date], value: Any) -> date | datetime:
+        date_obj = datetime.fromisoformat(value)
+        if issubclass(structure, datetime):
+            return date_obj
+        return date_obj.date()
+
+
+class StructureEncoder(EncoderProtocol, WithEncodeProtocol):
+    name: str
+    __type__ = (set, frozenset, Generator, Iterable, deque)
+
+    def serialize(self, obj: Any) -> list:
+        return list(cast(Any, obj))
+
+    def encode(self, structure: type, obj: Any) -> list:
+        return list(obj)
+
+
+DEFAULT_ENCODER_TYPES: deque[EncoderProtocol] = deque(
+    (
+        DataclassEncoder(),
+        NamedTupleEncoder(),
+        ModelDumpEncoder(),
+        EnumEncoder(),
+        PurePathEncoder(),
+        DateEncoder(),
+        StructureEncoder(),
+    )
+)
+
+ENCODER_TYPES: ContextVar[Sequence[EncoderProtocol]] = ContextVar(
+    "ENCODER_TYPES", default=DEFAULT_ENCODER_TYPES
+)
+
+
+def get_encoder_name(encoder: EncoderProtocol) -> str:
+    if getattr(encoder, "name", None):
+        return cast(str, encoder.name)
+    else:
+        return type(encoder).__name__
+
+
+def register_encoder(encoder: EncoderProtocol | type[EncoderProtocol]) -> None:
+    if isclass(encoder):
+        encoder = encoder()
+    if not isinstance(encoder, EncoderProtocol):
+        raise RuntimeError(f'"{encoder}" is not implementing the EncoderProtocol.')
+
+    encoder_name = get_encoder_name(encoder)
+    encoder_types = cast(deque[EncoderProtocol], ENCODER_TYPES.get())
+
+    remove_elements: list[EncoderProtocol] = []
+    for value in encoder_types:
+        if get_encoder_name(value) == encoder_name:
+            remove_elements.append(value)
+            break
+    for element in remove_elements:
+        encoder_types.remove(element)
+    encoder_types.appendleft(cast(Encoder[Any], encoder))
+
+
+def json_encoder_default(
+    value: Any,
+) -> Any:
+    """
+    Encode a value to a JSON-compatible format using a list of encoder types.
+
+    Parameters:
+    value (Any): The value to encode.
+
+    Returns:
+    Any: The JSON-compatible encoded value.
+
+    Raises:
+    ValueError: If the value is not serializable by any provided encoder type.
+    """
+    encoder_types = ENCODER_TYPES.get()
+
+    for encoder in encoder_types:
+        if encoder.is_type(value):
+            return encoder.serialize(value)
+
+    # If no encoder was found, raise a ValueError
+    raise ValueError(f"Object of type '{type(value).__name__}' is not JSON serializable.")
+
+
+def json_encoder(
+    value: Any,
+    *,
+    json_encode_fn: Callable[..., Any] = json.dumps,
```

Comment:
Missing docstrings for the remaining parameters and what they actually do and represent